agent/config/benchmark/uperf.json

```diff
@@ -0,0 +1,71 @@
+{
+  "controller" : {
+    "pre-script" : "/opt/pbench-agent/bench-scripts/postprocess/uperf-prepare-xmlfile",
+    "post-script" : "/opt/pbench-agent/bench-scripts/postprocess/uperf-postprocess-cdm",
+    "parameters" : { 
+      "defaults": {
+        "basic" : [ "--test-type=stream", "--protocol=tcp", "--message-size=16384",
+          "--instances=1", "--runtime=30"
+          ],
+        "full" : [ "--test-type=stream,maerts", "--protocol=tcp,udp",
+          "--message-size=64,256,1024,4096,16384,32768", "--instances=1,8",
+          "--runtime=30"
+          ],
+        "mandatory" : []
+      },
+      "specs" : {
+        "size_KMG" : {
+          "description" : "bytes in k/K (1024), m/M (1024^2), g/G (1024^3) or no units: 256 4k 16M 1g",
+          "arguments" : [ "message-size", "wsize", "rsize" ],
+       	  "value_regex" : "[0-9]+[kbmgKBZZMG]{0,1}",
+       	  "value_transforms" : [ "s/([0-9]+)[gG]/($1*1024).\"M\"/e",
+            "s/([0-9]+)[mM]/($1*1024).\"K\"/e", "s/([0-9]+)[kK]/($1*1024)/e"
+            ]
+          },
+        "hosts-or-IPs" : {
+          "description" : "list of valid hostnames or IPs",
+          "arguments" : [ "remotehost" ],
+          "value_regex" : "\\S+"
+          },
+        "test-types" : {
+          "description" : "all possible uperf test types",
+          "arguments" : [ "test-type" ],
+          "value_regex" : "^stream$|^maerts$|^rr$"
+          },
+        "protocols" : {
+          "description" : "all possible protocols",
+          "arguments" : [ "protocol" ],
+          "value_regex" : "^tcp|udp$"
+          },
+        "time_smh" : {
+          "description" : "time in seconds, minutes, or hours: 10s 2m 1h",
+          "arguments" : [ "runtime" ],
+          "value_regex" : "^[0-9]+[smh]{0,1}$",
+       	  "value_transforms" : [ "s/([0-9]+)[hH]/($1*60).\"M\"/e",
+            "s/([0-9]+)[mM]/($1*60)/e"
+            ]
+          },
+        "positive_integer" : {
+          "description" : "a whole number greater than 0",
+          "arguments" : [ "instances" ],
+          "value_regex" : "[1-9][0-9]*"
+          }
+        }
+      }
+    },
+  "client" : { 
+    "packages" : [ "pbench-uperf-1.0.4-1" ],
+    "pre-script" : " ",
+    "post-script" : " ",
+    "copy-to" : [ "/var/lib/pbench-agent/tmp/uperf.xml" ],
+    "bin" : "/opt/pbench-agent/bench-scripts/postprocess/uperf-client-launcher",
+    "param_regex" : []
+  },
+  "server" : { 
+    "packages" : [ "pbench-uperf-1.0.4-1" ],
+    "pre-script" : " ",
+    "post-script" : " ",
+    "bin" : "/opt/pbench-agent/bench-scripts/postprocess/uperf-server-launcher",
```

Comment:
Can we just assume all `bin` keys have values that are relative to the `pbench_install_dir` instead of hard-coding the paths?  Same for the `pre|post-script` keys above?

Reply:
I think we had a vision here where it would be possible for users to provide their own benchmark configuration files (ie. before something is upstream in pbench) so being able to reference ad hoc locations would be required.  In extreme cases this would allow someone to provide an implementation of a benchmark that we don't support ourselves, but in other cases it may just streamline the process of updating what we do support.

For example, for trafficgen we would do a normal (fully integrated) implementation for pbench-run-benchmark in pbench.  In parallel, we would have the source of those files included in our actual trafficgen repository (ie. the config JSON, the postprocessor, etc.).  When we update trafficgen (usually asynchronously to pbench updates) we would update the files in the trafficgen repo so that a user could take advantage of whatever the changes are prior to getting an updated version of pbench.  Once those updates are in trafficgen they would be synchronized (most likely with path updates) into pbench for a future pbench release.

If we follow this model it pushes the maintenance of these files (which are heavily tied to the benchmark) into the benchmark itself.

Reply:
It cannot be relative to pbench_install_dir because the "bin" could be literally anything, and is sometimes the actual benchmark binary itself, like fio.  If a benchmark absolutely has to have a wrapper script for the client or server, it will likely reside somewhere under pbench_install_dir.  To fix absolute paths that could eventually be wrong (like pbench_install_dir later changes to somethign other than /opt/pbenc-agent), we have a issue (https://github.com/distributed-system-analysis/pbench/issues/1320) in the run-benchmark project to allow the use of common pbench vars in the benchmark.json, so, for the uperf example, one could use: bin: $pbench_install_dir/postprocess/uperf-server-launcher.  For fio, since it does not use a wrapper script, should have either bin: "fio" or bin: "/usr/local/bin/fio" if we are certain that pbench-fio installed to /usr/local/bin/

Reply:
So then this begs the question of the location of the "launchers".  Seems like they are specific to uperf, so perhaps they should be in a separate location from the `post-process` directory to indicate they are uperf specific?  Perhaps a "contrib" area?

Perhaps we should have the `pbench-uperf` package provide those interfaces as an alternative so that the pbench-agent does not contain any benchmark specific mechanisms for `run-benchmark`.

In general this seems confusing to me and the interfaces seem tied directly to the implementation of `uperf` and `fio`.  Perhaps if we had an example "generic" benchmark to highlight the pbench-run-benchmark related interfaces separate from those of `uperf` and `fio` implementations that would make this clearer?

Reply:
Contrib in my mind would indicate this is optional, which it is not.  We don't keep $benchmark-postprocess files with pbench-$benchmark.RPM, so it does not make sense to include other benchmark helper scripts in those RPMs.  Those RPMs IMO are for the underlying benchmark only and nothing else.  Placing these scripts in the postprocess directory is probably not the most ideal place, but it was a convenient one.  They should go in a pbench directory that is not in the path (not bench-scripts and not tool-scripts), but should be part of pbench-agent because they are essential to running the benchmark.  Perhaps they should go in $pbench_install_dir/util-scripts/.

If we were following a more conforming installation, we would probably be placing almost all of this in /usr/libexec, and have the [eventually with pbench-run-benchmark] very few user-runnable scripts in /usr/bin/

Reply:
I'm going to work on implementing some other simple workloads this week just to get some more stuff in there.  I feel like we tackled three of the more complicated workloads (fio, uperf, trafficgen) first which was really good for making sure we covered the various nuances of a complicated workload but it does seem like it has made the barrier to entry, or understanding, a bit high.

Right now I'm thinking things like kernbench and some RT workloads which are pretty simplistic in nature, although the devil is in the details as usual.